### PR TITLE
Fix Issues #20, #40, and #45 by using the same escaping as Protobuf d…

### PR DESCRIPTION
**Pull request by mikegunter**
…istribution

The Unicode escaping that the Protocol Buffer distribution supports is
not implemented for now.

Add tests.

### CODE REVIEW

File: proto-lens-tests/tests/text_format_test.hs [hunk 1]
```diff
@@ -57,8 +65,40 @@ main = testMain
...
+          -- '\o172' == '\x7a' == 'z'
+         (Just $ def2 & b .~ "\o1\o12\o123\x2\o172z3z3")
+         (Data.Text.Lazy.pack "b: \"\\001\\012\\123\\002\\172\\x7a3\\1723\"")
+    , testCase "Render string with escape sequences" $
+        escapeRendered @=? showMessageShort escapeMessage
+    , readFrom "Parse rendered string with escape sequences"
+               (Just escapeMessage) (Data.Text.Lazy.pack escapeRendered)
+    , testCase "Render bytes" $
+         invalidUTF8BytesRendered @=? showMessage invalidUTF8BytesMessage
+    , readFrom "Parse single-quote-delimited string"
```
Comment (judah) [inline proto-lens-tests/tests/text_format_test.hs]: Can you add a test for an escape quote nested inside a string (assuming we don't already have it)?  For example, "abc\"def"

File: proto-lens/src/Data/ProtoLens/TextFormat.hs [hunk 2]
```diff
@@ -102,11 +104,35 @@ pprintFieldValue name SFixed64Field x = primField name x
...
+-- non-ASCII characters, which depending on the locale, the distribution might
+-- not do.
+--
+-- This uses three-digit octal escapes, e.g. "\011" plus \n, \r,, \t, \', \",
+-- and \\ only.  Note that Haskell string-literal syntax calls for "\011" to be
+-- interpreted as decimal 11, rather than the decimal 9 it actually represent,
+-- so you can't use Prelude.read to parse the strings created here.
+pprintByteString :: String -> Data.ByteString.ByteString -> Doc
+pprintByteString name x = text name <> colon <+> char '\"'
+    <> text (concatMap escape $ Data.ByteString.unpack x) <> char '\"'
```
Comment (judah) [inline proto-lens/src/Data/ProtoLens/TextFormat.hs]: Include the other ascii escapes like '\a', '\b', etc.?

Comment (mikegunter) [inline proto-lens/src/Data/ProtoLens/TextFormat.hs]: I followed the distribution, see https://github.com/google/protobuf/blob/master/src/google/protobuf/stubs/strutil.cc#L578, in not emitting those.

This inspired me to add a test for parsing them, though.

**Comment by mikegunter** [comment]
See the contents of escapeMessage and escapeRendered.

-m

**Pull request merged**